### DISCUSSION TITLE
Thinking about a special intermediate descriptive text format for some purposes

### DISCUSSION THREAD

**Discussion by Nakilon** [libvips/libvips#2825] [Ideas]
Two years ago using ruby-vips I was rendering maps for one video game.

![image](https://user-images.githubusercontent.com/2870363/170817993-f949f10b-188e-4496-88b4-fe79b54a3be1.png)

I kept things reproducible, i.e. there are stored background images taken from different websites and the project itself is a Ruby code that processes the game data files and then puts the results (texts and geometry primitives) onto those images. Also for my desired output some background image needed to be cropped, padded, color-adjusted, etc. so there was an "import config" -- a Ruby script with hardcoded transformations, it looked like this:

```ruby
      when "l08u_brainlab"
        loaded.embed(0, 0, loaded.width, loaded.height + 30, background: loaded.shrink(loaded.width, loaded.height).getpoint(0, 0))
      when "l12u_control_monolith"
        loaded.embed(20, 80, loaded.width + 40, loaded.height + 100).resize 0.5, vscale: 0.5, kernel: :lanczos2
      when "l10u_bunker"
        loaded.embed(0, 0, loaded.width, loaded.height, background: [0, 0, 0]).resize 1, vscale: 1, kernel: :lanczos2
```

During two weeks I have rendered "the same thing" a hundred of times while polishing the output on different stages of the process. I was on the edge of making some sort of cache, but how would you do it? You could just save to the `.v` files and wrap your code in some file-existence checking functions, but you would need to make up and remember the format of how you name the files, and to know where to put the wrappers you have to figure out which steps of the processing are related to the vips and which are not.

Imagine making it more automatic and universal. Imagine the whole "image" to be a chain of actions, like "take the file with this checksum and do these operations" -- this chain could be a key in the key-value cache. I didn't make it that time, it remained to be just an idea. I'll quote myself from the email:

> ... if to repetitively experiment with step 3 I don't want to wait again and again few seconds for steps 1 and 2 to finish it would be nice to somehow store those call args. This looks kind of like SVG data, that describes graphics in a textual way that you can process after the actual data generated but before it's rasterized. Like an imaginary "vips" data format that consists of call args specified by vips API.

Today I have this input image (from hairstoncreekfarm.com):

![image](https://user-images.githubusercontent.com/2870363/170817485-69ecd993-719e-4670-879d-3d20e1447ba9.png)

I want to preserve the original and I'm writing an "import config" right now:

```ruby
i = Vips::Image.new_from_file "Characteristics-of-Goose.jpg"
i.crop(30, 30, 160, 240).mutate do |i|
  10.step(i.width, 10){ |x| i.draw_line! 0,x,0,x,i.height }
  10.step(i.height, 10){ |y| i.draw_line! 0,0,y,i.width,y }
end.write_to_file "temp.png"
```

To extract each animal from there like this:

![image](https://user-images.githubusercontent.com/2870363/170817912-fdaa8918-31b7-46bc-9d69-d2d0f93d4897.png)

Now imagine if I call it a "vips script" and store it as a yaml:

```yaml
---
- - :new_from_file
  - Characteristics-of-Goose.jpg
- - :mutate
  - - :draw_image!
    - [0, 0, 0, 0, 872]
  - - :draw_image!
    - [0, 50, 0, 50, 872]
  ...
```

Won't it be convenient? It's not limited by the programming language and it's reproducible. So I see like two purposes of this, maybe someone sees more or has made something similar.

P.S.: how do I draw a semi-transparent line?

**Comment by jcupitt** [comment]
I was thinking about extending the `vips` command-line program to allow trees of operations, which I think would be similar functionality.

`vips` currently lets you give arguments to load and save operations, eg.:

```
vips copy infile.jpg[subsample=4] out.png[interlace]
```

That'll run `jpegload` and pass in `subsample=4`.

It'd be easy to extend that to any operation, not just loaders and savers. Something like:

```
vips invert {join x.jpg y.jpg} out.jpg`
```

ie. using curly brackets (maybe?) to indicate a new tree level. You could have things other than images too, perhaps:

```
vips relational_const x.jpg out.jpg more {avg x.jpg}
```

To find pixels greater than the average.

**Comment by jcupitt** [comment]
Semi-transparent lines: sorry, not possible directly. You could draw to an overlay and the composite that on top.

I think we'd need a `mode` option on `draw_line` like the one on `draw_image`.
